client/js/core/protocol.js

```diff
@@ -6,38 +6,168 @@ var Health = require("entities/model/health");
 var BioticState = require("entities/model/bioticState");
 var WorldItemState = require("entities/model/worldItemState");
 
-var Protocol = {};
+var Protocol = {
+	MSG_HANDSHAKE_REPLY:    0,
+	MSG_HANDSHAKE_ERROR:    1,
+	MSG_ENTITY_CREATE:      2,
+	MSG_ENTITY_STATE:       3,
+	MSG_ENTITY_REMOVE:      4,
+	MSG_CHUNK:              5,
+	MSG_BLOCK:              6,
+	MSG_CONTROLS_STATE:     7,
+	MSG_CHAT:               8,
+	MSG_DEBUG_RAY:          9,
+	MSG_NTP_SYNC:          10,
+	MSG_INVENTORY_STATE:   11,
+	MSG_INVENTORY_MOVE:    12,
+	MSG_SCOREBOARD_ADD:    13,
+	MSG_SCOREBOARD_SET:    14,
+	MSG_SCOREBOARD_REMOVE: 15
+};
+
+Protocol.append = function(a, b) {
+	var result = new Uint8Array(a.byteLength + b.byteLength);
+	result.set(new Uint8Array(a), 0);
+	result.set(new Uint8Array(b), a.byteLength);
+	return result.buffer;
+};
 
-Protocol.parseBody = function(proto) {
-	return new Body(
-		threeVecFromProto(proto.Pos),
-		threeVecFromProto(proto.Vel),
-		threeVecFromProto(proto.Dir),
-		threeVecFromProto(proto.HalfExtents),
-		threeVecFromProto(proto.CenterOffset)
-	)
+// Ported from varint.go
+Protocol.marshalInt = function(x) {
+	var buf = new Uint8Array(10);
+	var ux = (x << 1) >>> 0;
+	if (x < 0) {
+		ux = ~ux;
+	}
+
+	var i = 0;
+	while (ux >= 0x80) {
+		buf[i++] = ux | 0x80;
+		ux >>>= 7;
+	}
+	buf[i++] = ux;
+	return buf.buffer.slice(0, i);
 };
 
-Protocol.parseEntityState = function(proto) {
-	return new EntityState(proto.EntityId, Protocol.parseBody(proto.Body), proto.LastUpdated);
+// Ported from varint.go
+Protocol.unmarshalInt = function(offset, dataView) {
+	function returnValue(ux, n) {
+		var x = ux >>> 1;
+		if (ux&1 !== 0) {
+			x = ~x;
+		}
+		return { value: x, read: n };
+	}
+
+	var x = 0;
+	var s = 0;
+	var b;
+	for (var i = 0; ; i++) {
+		b = dataView.getUint8(offset + i);
+		if (b < 0x80) {
+			if (i > 9 || i === 9 && b > 1) {
+				return returnValue(0, -(i + 1)); // overflow
+			}
+			return returnValue(x | (b << s) >>> 0, i + 1);
+		}
+		x |= ((b&0x7F) << s) >>> 0;
+		s += 7;
+	}
```

Comment:
I've done a lot of testing myself, so if it doesn't change it's okay. But yeah, it would be nice to have some test mode to run and verify everything (particularly strings) bounce back and forth correctly.